DataFormats/Detectors/FIT/FT0/include/DataFormatsFT0/RawEventData.h

```diff
@@ -0,0 +1,218 @@
+        return;
+      }
+      printf("RDH| Ver:%2u Hsz:%2u Blgt:%4u FEEId:0x%04x PBit:%u\n",
+             uint32_t(h->version), uint32_t(h->headerSize), uint32_t(h->blockLength), uint32_t(h->feeId), uint32_t(h->priority));
+      printf("RDH|[CRU: Offs:%5u Msz:%4u LnkId:0x%02x Packet:%3u CRUId:0x%04x]\n",
+             uint32_t(h->offsetToNext), uint32_t(h->memorySize), uint32_t(h->linkID), uint32_t(h->packetCounter), uint32_t(h->cruID));
+      printf("RDH| TrgOrb:%9u HBOrb:%9u TrgBC:%4u HBBC:%4u TrgType:%u\n",
+             uint32_t(h->triggerOrbit), uint32_t(h->heartbeatOrbit), uint32_t(h->triggerBC), uint32_t(h->heartbeatBC),
+             uint32_t(h->triggerType));
+      printf("RDH| DetField:0x%05x Par:0x%04x Stop:0x%04x PageCnt:%5u\n",
+             uint32_t(h->detectorField), uint32_t(h->par), uint32_t(h->stop), uint32_t(h->pageCnt));
+    }
+  }
+  std::vector<char> to_vector()
+  {
+    constexpr int CRUWordSize = 16;
+    const char padding[CRUWordSize] = {0};
+
+    std::vector<char> result(size() * CRUWordSize);
+    char* out = result.data();
+    std::memcpy(out, &mEventHeader, EventHeader::PayloadSize);
+    out += EventHeader::PayloadSize;
+    LOG(DEBUG) << "write header for " << (int)mEventHeader.nGBTWords << " orbit " << int(mEventHeader.orbit) << " bc " << int(mEventHeader.bc);
+    if (mIsPadded) {
+      out += CRUWordSize - EventHeader::PayloadSize;
+      LOG(INFO) << "padding header " << CRUWordSize - EventHeader::PayloadSize;
+    }
+    for (int i = 0; i < mEventHeader.nGBTWords; ++i) {
+      std::memcpy(out, &mEventData[2 * i], EventData::PayloadSize);
+      out += EventData::PayloadSize;
+      std::memcpy(out, &mEventData[2 * i + 1], EventData::PayloadSize);
+      out += EventData::PayloadSize;
+      if (mIsPadded) {
+        out += CRUWordSize - 2 * EventData::PayloadSize;
+        LOG(INFO) << "padding data " << CRUWordSize - 2 * EventData::PayloadSize;
+      }
+    }
+    return result;
+  }
+  void setIsPadded(bool isPadding128)
+  {
+    mIsPadded = isPadding128;
+  }
+
+ public:
+  EventHeader mEventHeader;
+  EventData mEventData[Nchannels_PM];
+  bool mIsPadded = true;
+  /////////////////////////////////////////////////
+  ClassDefNV(RawEventData, 1);
+};
+std::ostream& operator<<(std::ostream& stream, const RawEventData& data);
+
+class DataPageWriter
+{
+  std::vector<char> mBuffer;
+  int mNpacketsInBuffer = 0;
+  std::vector<std::vector<char>> mPages;
+  std::vector<int> mNpackets;
+  static constexpr int MAX_PAGE_SIZE = 8192;
+
+ public:
+  o2::header::RAWDataHeader mRDH;
+  void flush(std::ostream& str)
+  {
+    writePage();
+    mRDH.stop = 0;
+    for (int page = 0; page < int(mPages.size()); ++page) {
+      mRDH.memorySize = mPages[page].size() + mRDH.headerSize;
+      mRDH.offsetToNext = mRDH.memorySize;
+      mRDH.packetCounter = mNpackets[page];
+      RawEventData::printRDH(&mRDH);
+      str.write(reinterpret_cast<const char*>(&mRDH), sizeof(mRDH));
```

Comment:
Are you still writing an RDH for each RawEventData? Why? Did not we agree that we will store in a new format: https://indico.cern.ch/event/791294/contributions/3456115/attachments/1869145/3074971/wp6.pdf)

Reply:
I am writing RDH when orbit is changed or data size >8K according new format.

Reply:
I call here in Raw2Digits
   for (const auto& digit : *digArr) {
      intRecord = digit.getInteractionRecord();
      uint32_t current_orbit = intRecord.orbit;
      LOG(DEBUG) << "old orbit " << old_orbit << " new orbit " << current_orbit;
      if (old_orbit != current_orbit) {
        for (DataPageWriter& writer : mPages)
          writer.flush(mFileDest);
        for (int nlink = 0; nlink < NPMs; ++nlink)
          setRDH(mPages[nlink].mRDH, nlink, intRecord);
        old_orbit = current_orbit;
      }

Reply:
But the RDH will NOT be written for each orbit, it will be writtern for the HB Frames separation only, see the link on Filippo's slides.

Reply:
O, previous design was HBorbit, i followed it.  So now only 8K ? OK.
But look at  my plots with comparison of digits and raw reconstruction
![T0ACdigrec](https://user-images.githubusercontent.com/20040756/69495038-b4eb8680-0ec2-11ea-8fab-1dd03ba0f557.png)
![timemult](https://user-images.githubusercontent.com/20040756/69495040-be74ee80-0ec2-11ea-8816-8868da3cf4c3.png)

Reply:
The RDH with new HB orbit/bc should be written at the beginning of every HB and repeated with PageCounter incremented and StopBit=0x1 at its end. In-between, the RDH is written (repeated with PageCounter incremented) only if the payload reached 8KB.

Concerning the plots: should not results of digits and raw reco be fully identical?

Reply:
About plots: difference comes from 13 ps channel width, digits are continuous.
About HB: I write RDH in the beginning of HBorbit with stop bit =0 and repeated with stop beat=1 at the end, before new one - as our hardware guys said me.  Payload size between 2 HB  is less then 8k event for Pb-Pb. But I have special check - and write RDH when it happens.

Reply:
Concerning the plots: sorry, I don't not understand what do you mean: digits and the raw data extracted from them are supposed to be equivalent.

For the HBF: [here](https://github.com/AliceO2Group/AliceO2/blob/6cf8c5cfb69152909be4a0b7ec9e33ec760c77fe/Detectors/FIT/FT0/simulation/src/Digits2Raw.cxx#L93) I see that you check the orbit of the digit and write a new RDH if it changes. Instead, your converter has to have a parameter saying how many orbits are in the HBFrame (by default 256), count the orbit changes and write the RDH only once a "new" HB starts.

Since you are at CERN, we better discuss details tomorrow in vivo.

Reply:
OK, when tomorrow? Morning?
About  plot: in digits time is written in ns, for raw i convert ns to number of channel. Should digits be in  number of channel ? I want to show these plots in Forward presentation tomorrow.

Reply:
Yes, preferably before 1pm (any time from 9:15).

Eventually, the FT0 reconstruction should provide BC + time in ns, right? So I would expect this info obtained both from digits and raw data.

Reply:
the FT0 reconstruction  provides BC + time in ns.
9:30, OK?

Reply:
Yes, 9:30 is ok

Reply:
Hi @AllaMaevskaya 
sorry, I've completely messed up the comments above. The RDH should be indeed added for every HB, i.e. orbit. It is the TF trigger flag (bit 11 in the trigger) which should be set at each HB%255 == 0. But since the 1st HB phase and counting should be done in a same way for all detectors, we better have a common method for that, rather than just checking the orbit change. Also note that the orbit may jump by >1 (if there were no interactions in one orbi), still the empty HBF for the skipped orbit should be written.
Please, don't change at the moment your ``Digits2Raw.cxx``, I will provide a piece of code for common HB generation.